src/main/java/com/google/alpollo/servlets/AutocompleteServlet.java

```diff
@@ -0,0 +1,90 @@
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.alpollo.helpers.ConfigHelper;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.gson.Gson;
+import com.google.gson.JsonIOException;
+import com.google.gson.JsonSyntaxException;
+import com.jayway.jsonpath.JsonPath;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+@WebServlet("/autocomplete")
+public class AutocompleteServlet extends HttpServlet {
+  private final Gson gson = new Gson();
+  private final static String SEARCH_STRING = "searchString";
+  private final static String TYPE = "type";
+  private final static String LIMIT = "10";
+  private List<String> typeList;
+  private enum SearchType {
+    ARTIST,
+    SONG;
+  };
+
+  /**
+   * Making a POST request to this servlet with a search string and the type searched for
+   * as parameters will make a call to the Knowledge Graph Search API and it will return a list of
+   * possible search results.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String searchString;
+    SearchType type;
+    try {
+      searchString = request.getParameter(SEARCH_STRING);
+      type = SearchType.valueOf(request.getParameter(TYPE));
+      switch (type) {
+        case ARTIST:
+          typeList = Arrays.asList("Person", "MusicGroup");
+          break;
+        case SONG:
+          typeList = Arrays.asList("MusicRecording");
+          break;
+      }
+
+      HttpTransport httpTransport = new NetHttpTransport();
+      HttpRequestFactory requestFactory = httpTransport.createRequestFactory();
+
+      JSONParser parser = new JSONParser();
+
+      GenericUrl url = new GenericUrl("https://kgsearch.googleapis.com/v1/entities:search");
+      url.put("query", searchString);
+      url.put("limit", LIMIT);
+      url.put("types", typeList);
+      url.put("key", ConfigHelper.getSensitiveData(this.getServletContext(), ConfigHelper.SENSITIVE_DATA.API_KEY));
+
+      HttpRequest autocompleteRequest = requestFactory.buildGetRequest(url);
+      HttpResponse autocompleteResponse = autocompleteRequest.execute();
+      JSONObject responseObject = (JSONObject) parser.parse(autocompleteResponse.parseAsString());
+      JSONArray elements = (JSONArray) responseObject.get("itemListElement");
+
+      List<String> results = new ArrayList<>();
+      for (Object element : elements) {
+        results.add(JsonPath.read(element, "$.result.name").toString());
+      }
```

Comment:
Can this not done with Gson, instead of adding new dependencies?

Reply:
JsonPath from jayway apparently doesn't like to work with the jsons created by gson. This is what I'm getting if I use gson:
Expected to find an object with property ['result'] in path $ but found 'com.google.gson.JsonObject'. This is not a json object according to the JsonProvider: 'com.jayway.jsonpath.spi.json.JsonSmartJsonProvider'.

Reply:
It looks like you might be able to have JsonPath work with Gson by using a custom configuration: https://github.com/json-path/JsonPath#tweaking-configuration